proposals/lip-0051.md

```diff
@@ -321,178 +385,255 @@ The module provides the following commands to modify token entries.
+            "maxLength": MAX_DATA_LENGTH,
             "fieldNumber": 4
+        },
+        "accountInitializationFee": {
+            "dataType": "uint64",
+            "fieldNumber": 5
         }
     }
 }
 ```
 
-##### Parameters Validity
-
-The `params` property of a token transfer transaction is valid if:
-
-* `recipientAddress` is a byte array of length `ADDRESS_LENGTH`.
-* `data` has length less than or equal to `MAX_DATA_LENGTH`.
+The property `accountInitializationFee` is of interest only in cases where there is no entry `userStore(recipientAddress, tokenID)` and the sender needs to pay a fee for the creation of this entry. In such cases, it should have value `USER_ACCOUNT_INITIALIZATION_FEE`. The token for this fee is `TOKEN_ID_STORE_INITIALIZATION`; by default it is the token used for transaction fees in the chain, i.e., `TOKEN_ID_STORE_INITIALIZATION = TOKEN_ID_FEE`. In case the entry `userStore(recipientAddress, tokenID)` already exists and the sender of the transaction sets a non-zero amount for this property, then sender's account is not debited. 
 
-##### Execution
 
-When executing a token transfer transaction `trs`, the logic below is followed:
+##### Verification
 
 ```python
-derive senderAddress from trs.senderPublicKey
-let tokenID, recipientAddress, amount given by trs.params
+def verify(trs: Transaction):
+    senderAddress = SHA256(trs.senderPublicKey)[:ADDRESS_LENGTH]         # derive sender address from trs.senderPublicKey
+    tokenID = trs.params.tokenID
+    recipientAddress = trs.params.recipientAddress
+    amount = trs.params.amount 
+
+    
+    balanceChecks: dict[TokenID, uint64] = {}
+
+    if amount == 0:
+        raise Exception('Transfer amount should be non-zero.')
+    # check whether account initialization fee is relevant
+    if  userStore(recipientAddress, tokenID) does not exist:
+        if trs.params.accountInitializationFee != USER_ACCOUNT_INITIALIZATION_FEE:
+            raise Exception("Invalid account initialization fee")
+        
+        balanceChecks[TOKEN_ID_STORE_INITIALIZATION] = USER_ACCOUNT_INITIALIZATION_FEE
+
+    if tokenID in balanceChecks:
+        balanceChecks[tokenID] += amount
+    else:
+        balanceChecks[tokenID] = amount
+    
+    for tkID in balanceChecks:
+        if availableBalance(senderAddress, tkID) < balanceChecks[tkID]:
+            raise Exception("Insufficient balance")
+```
 
-if availableBalance(senderAddress, tokenID) < amount:
-    transaction execution fails
+##### Execution
 
-availableBalance(senderAddress, tokenID) -= amount
-availableBalance(recipientAddress, tokenID) += amount
+```python
+def execute(trs: Transaction):
+
+    senderAddress = SHA256(trs.senderPublicKey)[:ADDRESS_LENGTH]         # derive sender address from trs.senderPublicKey
+    tokenID = trs.params.tokenID
+    recipientAddress = trs.params.recipientAddress
+    amount = trs.params.amount
+    accountInitializationFee = trs.params.accountInitializationFee 
```

Comment:
this variable is not used

Reply:
Good catch, not used anymore in internal. Removed.